src/prefect/client/orion.py

```diff
@@ -975,8 +975,12 @@ async def create_block_document(
         except httpx.HTTPStatusError as e:
             if e.response.status_code == status.HTTP_409_CONFLICT:
                 raise prefect.exceptions.ObjectAlreadyExists(http_exc=e) from e
-            else:
-                raise
+            elif e.response.status_code == status.HTTP_422_UNPROCESSABLE_ENTITY:
+                for detail in e.response.json().get("exception_detail", []):
+                    message = detail.get("msg", "")
+                    if "name must only contain" in message:
+                        raise prefect.exceptions.InvalidNameError(message) from e
```

Comment:
Instead of this, we should consider not sending this request to the API at all since we know it's invalid. If you add a validator to `BlockDocumentCreate`, then the error will be thrown client-side during construction of the document.

Reply:
It seems like there already is a validator in BlockDocumentCreate:

```



def validate_block_document_name(value):
    if value is not None and not bool(
        re.match(LOWERCASE_LETTERS_AND_DASHES_ONLY_REGEX, value)
    ):
        raise ValueError(
            "name must only contain lowercase letters, numbers, and dashes"
        )
    return value
...
@copy_model_fields
class BlockDocumentCreate(ActionBaseModel):
    """Data used by the Orion API to create a block document."""

    name: Optional[str] = FieldFrom(schemas.core.BlockDocument)
    data: dict = FieldFrom(schemas.core.BlockDocument)
    block_schema_id: UUID = FieldFrom(schemas.core.BlockDocument)
    block_type_id: UUID = FieldFrom(schemas.core.BlockDocument)
    is_anonymous: bool = FieldFrom(schemas.core.BlockDocument)

    _validate_name_format = validator("name", allow_reuse=True)(
        validate_block_document_name
    )

    @root_validator
    def validate_name_is_present_if_not_anonymous(cls, values):
        # TODO: We should find an elegant way to reuse this logic from the origin model
        if not values.get("is_anonymous") and not values.get("name"):
            raise ValueError("Names must be provided for block documents.")
        return values
```

Reply:
I guess `_save`'s `_to_block_document` doesn't return `BlockDocumentCreate`:

```

        try:
            block_document = await client.create_block_document(
                block_document=self._to_block_document(name=name)
            )
```

```

    def _to_block_document(
        self,
        name: Optional[str] = None,
        block_schema_id: Optional[UUID] = None,
        block_type_id: Optional[UUID] = None,
        is_anonymous: Optional[bool] = None,
    ) -> BlockDocument:
```

Reply:
Sounds like we should pass the correct type to the client or the client should coerce it to the expected type.